### ISSUE TITLE
Document how SigningKeyResolver implementations should signal there's no matching key

### ISSUE THREAD

**Issue by bturner** [jwtk/jjwt#713]
**Is your feature request related to a problem? Please describe.**
When building an implementation for `SigningKeyResolver`, the Javadocs for the interface provide no hints as to how an implementation should react when it cannot find a key. Returning `null` is not handled by the calling code in JJWT, so it _appears_ the correct thing to do is throw an exception. It would be nice if the Javadocs simply stated explicitly what an implementor should do.

**Describe the solution you'd like**
Update the Javadocs to indicate how `SigningKeyResolver` implementations should react when there's no matching key.

**Additional context**
I ended up implementing my `SigningKeyResolver` to throw a `SignatureException`, since that seemed to be what the surrounding code in JJWT would do (for example, if a key was returned but could not verify the signature).

**Comment by bdemers** [comment]
Either throwing or returning `null` will work, (the resolved key being null is checked here: https://github.com/jwtk/jjwt/blob/6b2843bded4ae9c446b74cf2b12f104c5ca97b52/impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java#L390 )

I agree, making this more clear in the Javadoc would help. We can add something like "or {@code null} if a corresponding key could not be found"
https://github.com/jwtk/jjwt/blob/439c027/api/src/main/java/io/jsonwebtoken/SigningKeyResolver.java#L58-L69

I hint at `null`, because you _may_ want to aggregate resolvers together, but that gets into use case-specific details.

**Comment by lhazlewood** [comment]
Thanks @bturner for the issue! (@bdemers, Bryan and I go *way* back - he was the very first Stormpath engineer! 😎 )

And agreed, `Assert.notNull` will indeed throw an exception, but it is just an `IllegalArgumentException`, which is maybe not the best exception to self-document what is going on.  I suppose an `IllegalArgumentException` at the end of the day does work because the 'illegal argument' is the jwt token being parsed, so that I think can make sense, but it's maybe a technicality rather than what would be ideal.

In the JWE branch, a new `Locator<T>` concept is replacing the `SigningKeyResolver`  so perhaps a new Exception subclass would be in order. The replacement is because it needs to resolve potentially more than just signing/verification keys - if encountering a JWE, it also needs to locate decryption keys, so a more generic 'Locator' name has been chosen.  Basically anything that needs to be looked up based on a Header - e.g. compression algorithm, whatever - is a candidate implementation of a `Locator<T>`.  Here is what it looks like:

```java
public interface Locator<T> {
    T locate(Header<?> header);
}
```

The difficulty is figuring out what the new exception should subclass - `UnsupportedJwtException` (because something the header references isn't available), or a `KeyException` (e.g. `MissingKeyException`), or an entirely new `HeaderException` hierarchy, I dunno.

**Comment by bturner** [comment]
`HeaderException` doesn't feel like the right fit for this, to me, because there's nothing wrong with the headers (necessarily). I'd expect that to be more related to parsing than resolving keys. `NoSuchKeyException` or `UnknownKeyException` or `MissingKeyException`, something more indicative that the system tried to resolve a key but couldn't find one, that would be my vote.

Alternatively, if any  exception my resolver throws is simply going to be passed through as-is, rather than caught and wrapped, perhaps that's all the documentation would need to say--that you can throw an application-specific exception that your surrounding code (because we all write good `catch` blocks, right? :badpokerface:) can catch and handle however is appropriate. I only think a new exception might be needed if the plan is to document that the resolver can return `null` and JJWT's internals will automatically throw. (I think it'd be a nice touch if the new exception could include the `Header`, since those must have been parsed correctly for JJWT to try and look up a key.)

**Comment by lhazlewood** [comment]
Thanks for the insight Bryan - we can definitely add these suggestions.  I do really like the Header being included in the exception - that makes total sense.

Unfortunately, I didn't explain my thoughts around the `HeaderException` concept too well - that would really be an intermediate exception in a hierarchy from which all other header-related exceptions subclass.  It would parallel (in concept) the `ClaimJwtException` (which really should have just been named `ClaimException`) in the screenshot - individual subclasses explain exactly why something was wrong due to a claim.  

Similarly, a `HeaderException` would exist to indicate something was wrong with a header.  Not that they're malformed (that would be a `MalformedJwtException`, but that there's something semantically incorrect about one or more of them.

A subclass of `HeaderException`  - maybe like `UnknownReferenceException` (or anything similar/better) - could exist for "Your header is referencing something - like a compression algorithm, or key or public key, or whatever, and we couldn't find it" scenarios.  This is because there are more than just Keys that can be located via headers.

What's difficult to tell is should this generic concept ("I can't locate something needed") be ok to use for key lookups?

I'm _assuming_ so, because a resource lookup is an orthogonal concern to signature verification or decryption, so a `SignatureException` doesn't quite feel right for some reason IMO (we have no idea if anything was wrong with the signature or the verification process because we couldn't even get that far).

Another thought - A `HeaderException` hierarchy could ensure that it and all subclasses contain the `Header` instance as you suggested.  That's a little harder to enforce on other exceptions that are not in that sub-hierarchy (like KeyExceptions).

**Comment by lhazlewood** [comment]
Not stale, but it is probably resolved in the `jwe` branch with the addition of the new `Locator<Header>` interface, and that does indeed document that a null return value means 'cannot be found'

**Comment by lhazlewood** [comment]
Closing as this is resolved by the new `Locator<Header>` and `JwtParserBuilder.keyLocator` support merged recently into `master` as `SigningKeyResolver` is now deprecated.  This will be released in `0.12.0`.
